File: src/main/java/javaclasses/LruCache/LRUCache.java
package javaclasses.LruCache;

import java.util.HashMap;

/*Design a data structure that follows the constraints of a Least Recently Used (LRU) cache.

        Implement the LRUCache class:

        LRUCache(int capacity) Initialize the LRU cache with positive size capacity.
        int get(int key) Return the value of the key if the key exists, otherwise return -1.
        void put(int key, int value) Update the value of the key if the key exists. Otherwise, add the key-value pair to the cache.
        If the number of keys exceeds the capacity from this operation, evict the least recently used key.
        Follow up:
        Could you do get and put in O(1) time complexity?

        Example 1:

        Input
        ["LRUCache", "put", "put", "get", "put", "get", "put", "get", "get", "get"]
        [[2], [1, 1], [2, 2], [1], [3, 3], [2], [4, 4], [1], [3], [4]]
        Output
        [null, null, null, 1, null, -1, null, -1, 3, 4]

        Explanation
        LRUCache lRUCache = new LRUCache(2);
        lRUCache.put(1, 1); // cache is {1=1}
        lRUCache.put(2, 2); // cache is {1=1, 2=2}
        lRUCache.get(1);    // return 1
        lRUCache.put(3, 3); // LRU key was 2, evicts key 2, cache is {1=1, 3=3}
        lRUCache.get(2);    // returns -1 (not found)
        lRUCache.put(4, 4); // LRU key was 1, evicts key 1, cache is {4=4, 3=3}
        lRUCache.get(1);    // return -1 (not found)
        lRUCache.get(3);    // return 3
        lRUCache.get(4);    // return 4*/

//https://leetcode.com/problems/lru-cache/
public class LRUCache {
    Node head;
    Node tail;

    HashMap<Integer, Node> map = null;
    int cap = 0;

    public LRUCache(int capacity) {
        this.cap = capacity;
        this.map = new HashMap<>();
    }

    public void put(int key, int value) {
        if (map.containsKey(key)) {
            Node t = map.get(key);
            t.value = value;
            //move to tail
            removeNode(t);
            insertNode(t);
        } else {
            if (map.size() >= cap) {
                //delete head
                map.remove(head.key);
                removeNode(head);
            }
            //add to tail
            Node node = new Node(key, value);
            insertNode(node);
            map.put(key, node);
        }
    }

    public int get(int key) {
        if (map.get(key) == null) {
            return -1;
        }
        //mode to tail
        Node t = map.get(key);
        removeNode(t);
        insertNode(t);
        return t.value;
    }

    public void removeNode(Node n) {
        if (n.prev != null) {
            n.prev.next = n.next;
        } else {
            head = n.next;
        }

        if (n.next != null) {
            n.next.prev = n.prev;
        } else {
            tail = n.prev;
        }
    }

    public void insertNode(Node n) {
        if (tail != null) {
            tail.next = n;
        }
        n.prev = tail;
        n.next = null;
        tail = n;
        if (head == null) {
            head = tail;
        }
    }
}

File: src/main/java/kotlinclasses/DataStructure/LRUCache.kt
/*
 * *
 *  * LRU Cache.kt
 *  * Created by Rafsan Ahmad on 3/15/25, 7:56PM
 *  * Copyright (c) 2025 . All rights reserved.
 *  *
 *
 */

package kotlinclasses.DataStructure

class LRUCache(capacity: Int) {
    //https://leetcode.com/problems/lru-cache/description/
    class Node(var key: Int, var value: Int) {
        var prev: Node? = null
        var next: Node? = null
    }

    val map: MutableMap<Int, Node> = mutableMapOf() // Stores key to Node mapping for O(1) access
    var head: Node? = null  // Most recently used node
    var tail: Node? = null  // Least recently used node
    val cap = capacity      // Capacity limit of cache

    //Time Complexity: O(1)
    fun get(key: Int): Int {
        map[key]?.let {
            deleteNode(it)    // Remove node from its current position in the list
            insertNode(it)    // Insert node at head (mark as most recently used)
            return it.value   // Return the value associated with the key
        }
        return -1             // Key not found
    }

    //Time Complexity: O(1)
    fun put(key: Int, value: Int) {
        map[key]?.let {
            it.value = value  // Update existing node's value
            deleteNode(it)    // Remove node from its current position
            insertNode(it)    // Insert node at head (MRU)
            return
        }

        if (map.size >= cap) {
            tail?.let {
                map.remove(it.key)  // Remove least recently used key from map
                deleteNode(it)      // Remove least recently used node from list
            }
        }

        val newNode = Node(key, value) // Create new node for new key-value
        map[key] = newNode             // Add to map for O(1) access
        insertNode(newNode)            // Insert at head (most recently used)
    }

    //Time Complexity: O(1)
    fun insertNode(node: Node) {
        if (head == null) {
            head = node              // If list empty, new node becomes head
        } else {
            head?.prev = node       // Link current head prev to new node
            node.next = head        // New node points next to current head
            head = node             // Update head to new node
        }
        if (tail == null)
            tail = node             // If list was empty, tail is also new node
        node.prev = null            // IMPORTANT: New head always has null prev
    }

    //Time Complexity: O(1)
    fun deleteNode(node: Node) {
        if (node.prev == null) {
            // Node is head
            head = node.next
        } else {
            node.prev?.next = node.next // Link previous node's next to node's next
        }

        if (node.next == null) {
            // Node is tail
            tail = node.prev
        } else {
            node.next?.prev = node.prev // Link next node's prev to node's prev
        }
    }
}

fun main() {
    val lru = LRUCache(2)  // capacity 2

    lru.put(1, 1)          // cache = {1=1}
    lru.put(2, 2)          // cache = {1=1, 2=2}

    println(lru.get(1))    // returns 1, cache = {2=2, 1=1}

    lru.put(3, 3)          // evicts key 2, cache = {1=1, 3=3}

    println(lru.get(2))    // returns -1 (not found)

    lru.put(4, 4)          // evicts key 1, cache = {3=3, 4=4}

    println(lru.get(1))    // returns -1 (not found)
    println(lru.get(3))    // returns 3
    println(lru.get(4))    // returns 4
}